_posts/2022-05-18-torchopt/torchopt.Rmd

```diff
@@ -0,0 +1,356 @@
+
+[`torchopt`](https://github.com/e-sensing/torchopt) is a package developed by [Gilberto Camara](https://gilbertocamara.org/) and colleagues at [National Institute for Space Research, Brazil](https://www.gov.br/inpe/pt-br).
+
+By the look of it, the package's reason of being is rather self-evident. `torch` itself does not -- nor should it -- implement all the newly-published, potentially-useful-for-your-purposes optimization algorithms out there.[^2] The algorithms assembled here, then, are probably exactly those the authors were most eager to experiment with in their own work. As of this writing, they comprise, amongst others, various members of the popular *ADA*\* and \**ADAM*\* families.[^3] And we may safely assume the list will grow over time.
+
+[^2]: For why it shouldn't, and an overview of (some of) the things `torch` users can do, see also our recent post on [extending `torch`](https://blogs.rstudio.com/ai/posts/2022-04-27-torch-outside-the-box/).
+
+[^3]: See the [Readme](https://github.com/e-sensing/torchopt) for an up-to-date listing.
+
+I'm going to introduce the package by highlighting something that technically, is "merely" a utility function, but to the user, can be extremely helpful: the ability to, for an arbitrary optimizer and an arbitrary test function, plot the steps taken in optimization.
+
+While it's true that I have no intent of comparing (let alone analyzing) different strategies, there is one that, to me, stands out in the list: ADAHESSIAN [@abs-2006-00719], a second-order algorithm designed to scale to large neural networks. I'm especially curious to see how it behaves as compared to L-BFGS, the second-order "classic" available from base `torch` we've had a [dedicated blog post](https://blogs.rstudio.com/ai/posts/2021-04-22-torch-for-optimization/) about last year.
+
+## The way it works
+
+The utility function in question is named `test_optim()`. The only required argument concerns the optimizer to try (`optim`). But you'll likely want to tweak three others as well:
+
+-   `test_fn`: To use a test function different from the default (`beale`). You can choose among the many provided in `torchopt`, or you can pass in your own. In the latter case, you also need to provide information about search domain and starting points. (We'll see that in an instant.)
+-   `steps`: To set the number of optimization steps.
+-   `opt_hparams`: To modify optimizer hyperparameters; most notably, the learning rate.
+
+Here, I'm going to use the `flower()` function that already prominently figured in the aforementioned post on [L-BFGS](https://blogs.rstudio.com/ai/posts/2021-04-22-torch-for-optimization/). It approaches its minimum as it gets closer and closer to `(0,0)` (but is undefined at the origin itself).
+
+Here it is:
+
+```{r}
+flower <- function(x, y) {
+  a <- 1
+  b <- 1
+  c <- 4
+  a * torch_sqrt(torch_square(x) + torch_square(y)) + b * torch_sin(c * torch_atan2(y, x))
+}
+```
+
+Let's begin with our explorations.
```

Comment:
Would it make sense to show a plot of the flower function here, and maybe explain a little how to read that plot?

Reply:
Yeah, I can see what you mean... I was intentionally not doing it, because of the many plots  to come ...but then, it doesn't hurt, really ;-))
And true, makes sense to shortly explain the plot, especially since - as I noticed - the color coding works exactly the opposite way from what I was doing for the other post!